File: repos/mio/by-name/ha/hanga-contrib/lib/go/hangamod/wire.go
package hangamod

type WireKind uint8

const (
	WireEmpty WireKind = iota
	WireFlag
	WireInt
	WireFloat
	WireText
	WireList
	WireBag
	WireFail
)

type Field struct {
	Key   string
	Value Wire
}

// JSON-shaped payload: null, bool, number, string, array, object.
type Wire struct {
	Kind   WireKind
	Flag   bool
	Int    int64
	Float  float64
	Text   string
	Items  []Wire
	Fields []Field
}

func TextWire(value string) Wire {
	return Wire{Kind: WireText, Text: value}
}

func FailWire(reason string) Wire {
	return Wire{Kind: WireFail, Text: reason}
}

func VoxelProbe(name string, edit bool) Wire {
	return Wire{
		Kind: WireBag,
		Fields: []Field{
			{Key: "name", Value: Wire{Kind: WireText, Text: name}},
			{Key: "edit", Value: Wire{Kind: WireFlag, Flag: edit}},
		},
	}
}

func (w Wire) AsText() (string, bool) {
	if w.Kind != WireText {
		return "", false
	}
	return w.Text, true
}

func (w Wire) BagText(key string) (string, bool) {
	if w.Kind != WireBag {
		return "", false
	}
	for _, field := range w.Fields {
		if field.Key == key && field.Value.Kind == WireText {
			return field.Value.Text, true
		}
	}
	return "", false
}

func (w Wire) BagFlag(key string) bool {
	if w.Kind != WireBag {
		return false
	}
	for _, field := range w.Fields {
		if field.Key != key {
			continue
		}
		switch field.Value.Kind {
		case WireFlag:
			return field.Value.Flag
		case WireInt:
			return field.Value.Int == 1
		}
	}
	return false
}

func (w Wire) BagInt(key string) int64 {
	if w.Kind != WireBag {
		return 0
	}
	for _, field := range w.Fields {
		if field.Key != key {
			continue
		}
		switch field.Value.Kind {
		case WireInt:
			return field.Value.Int
		case WireText:
			n, _ := parseInt64(field.Value.Text)
			return n
		}
	}
	return 0
}

File: repos/mio/by-name/ha/hanga-contrib/lib/nim/hangamod/wire.nim
import strutils

type
  WireKind* = enum
    wkEmpty, wkFlag, wkInt, wkFloat, wkText, wkList, wkBag, wkFail

  Field* = object
    key*: string
    value*: Wire

  Wire* = object
    kind*: WireKind
    flag*: bool
    intVal*: int64
    floatVal*: float64
    text*: string
    items*: seq[Wire]
    fields*: seq[Field]

proc textWire*(value: string): Wire =
  Wire(kind: wkText, text: value)

proc failWire*(reason: string): Wire =
  Wire(kind: wkFail, text: reason)

proc voxelProbe*(name: string, edit: bool): Wire =
  Wire(
    kind: wkBag,
    fields: @[
      Field(key: "name", value: Wire(kind: wkText, text: name)),
      Field(key: "edit", value: Wire(kind: wkFlag, flag: edit)),
    ],
  )

proc asText*(w: Wire): tuple[ok: bool, value: string] =
  if w.kind == wkText:
    (true, w.text)
  else:
    (false, "")

proc bagText*(w: Wire, key: string): tuple[ok: bool, value: string] =
  if w.kind != wkBag:
    return (false, "")
  for field in w.fields:
    if field.key == key and field.value.kind == wkText:
      return (true, field.value.text)
  (false, "")

proc bagFlag*(w: Wire, key: string): bool =
  if w.kind != wkBag:
    return false
  for field in w.fields:
    if field.key != key:
      continue
    case field.value.kind
    of wkFlag:
      return field.value.flag
    of wkInt:
      return field.value.intVal == 1
    else:
      discard
  false

proc bagInt*(w: Wire, key: string): int64 =
  if w.kind != wkBag:
    return 0
  for field in w.fields:
    if field.key != key:
      continue
    case field.value.kind
    of wkInt:
      return field.value.intVal
    of wkText:
      try:
        return parseBiggestInt(field.value.text)
      except CatchableError:
        return 0
    else:
      discard
  0
